subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionEngine.java

```diff
@@ -45,4 +52,31 @@
      * The work is looked up by its {@link UnitOfWork.Identity identity} in the given cache.
      */
     <T, O> T getFromIdentityCacheOrDeferExecution(UnitOfWork work, Cache<Identity, Try<O>> cache, DeferredExecutionHandler<O, T> handler);
+
+    interface Result {
+        Try<ExecutionResult> getExecutionResult();
+
+        CachingState getCachingState();
+
+        /**
+         * A list of messages describing the first few reasons encountered that caused the work to be executed.
+         * An empty list means the work was up-to-date and hasn't been executed.
+         */
+        ImmutableList<String> getExecutionReasons();
+
+        /**
+         * If a previously produced output was reused in some way, the reused output's origin metadata is returned.
+         */
+        Optional<OriginMetadata> getReusedOutputOriginMetadata();
+
+        /**
+         * Snapshots of the roots of output properties.
+         *
+         * In the presence of overlapping outputs this might be different from
+         * {@link org.gradle.internal.execution.history.BeforeExecutionState#getOutputFileLocationSnapshots()},
```

Comment:
Does the public API now reference some internal API via the Javadoc?

Reply:
Good point, should be fixed.